compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -291,7 +324,51 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             } else {
                 false
             };
-            param_count_error(expected_arg_count, supplied_arg_count, "E0061", false, sugg_unit);
+
+            // are we passing elements of a tuple without the tuple parentheses?
+            let expected_arg_types = if expected_arg_tys.is_empty() {
+                // In most cases we can use expected_arg_tys, but some callers won't have the type
+                // information, in which case we fall back to the types from the input expressions.
+                fn_inputs
+            } else {
+                &expected_arg_tys
+            };
+
+            let sugg_tuple_wrap_args = if let [expected_arg_type] = expected_arg_types[..] {
+                let expected_arg_type = self.resolve_vars_if_possible(expected_arg_type);
+
+                if let ty::Tuple(expected_elems) = expected_arg_type.kind() {
+                    let expected_types: Vec<_> =
+                        expected_elems.iter().map(|k| k.expect_ty()).collect();
+                    let supplied_types: Vec<_> =
+                        args.iter().map(|arg| self.check_expr(arg)).collect();
+
+                    if iter::zip(expected_types, supplied_types).all(|(expected, supplied)| {
+                        self.can_eq(ParamEnv::empty(), expected, supplied).is_ok()
+                    }) {
+                        match args {
+                            [] => None,
+                            [single] => Some(FnArgsAsTuple::Single(single)),
+                            [first, .., last] => Some(FnArgsAsTuple::Multi { first, last }),
+                        }
+                    } else {
```

Comment:
I'm not particularly proud of these `else { None }` flourishes at the end here, although I think it's perhaps more readable than the equivalent functional chain - any thoughts?

Reply:
I would suggest extracting a helper function for this code. That always helps readability.

Additionally, once you extract a helper function, you can use the new `let_else` feature to replace these nested `if let`s with guarded returns. E.g.:

```rust
let [expected_arg_type] = expected_arg_types[..]
else { return None };
```

and you could convert the plain `if` to a guarded return as well.

Reply:
Was totally waiting for that feature to be in stable, forgot we can do this kinda thing in `rustc` - thanks!